cpp/open3d/t/pipelines/registration/Registration.cpp

```diff
@@ -367,64 +341,64 @@ RegistrationResult MultiScaleICP(
-                    estimation, num_iterations);
+                    max_correspondence_distances[num_scales - 1], estimation,
+                    num_scales);
 
     // Transformation tensor is always of shape {4,4}, type Float64 on CPU:0.
     core::Tensor transformation =
             init_source_to_target.To(core::Device("CPU:0"), core::Float64);
-    RegistrationResult result(transformation, save_loss_log);
-
-    double prev_fitness = 0;
-    double prev_inlier_rmse = 0;
+    RegistrationResult result(transformation);
 
+    int iteration_count = 0;
     // ---- Iterating over different resolution scale START -------------------
-    for (int64_t i = 0; i < num_iterations; ++i) {
-        source_down_pyramid[i].Transform(transformation);
-
+    for (int64_t scale_idx = 0; scale_idx < num_scales; ++scale_idx) {
+        source_down_pyramid[scale_idx].Transform(result.transformation_);
         // Initialize Neighbor Search.
         core::nns::NearestNeighborSearch target_nns(
-                target_down_pyramid[i].GetPointPositions());
-        bool check = target_nns.HybridIndex(max_correspondence_distances[i]);
+                target_down_pyramid[scale_idx].GetPointPositions());
+        bool check =
+                target_nns.HybridIndex(max_correspondence_distances[scale_idx]);
         if (!check) {
-            utility::LogError(
-                    "NearestNeighborSearch::HybridSearch: Index is not set.");
+            utility::LogError("Index is not set.");
         }
 
         // ICP iterations result for single scale.
-        DoSingleScaleIterationsICP(
-                source_down_pyramid[i], target_down_pyramid[i], target_nns,
-                criterias[i], max_correspondence_distances[i], transformation,
-                estimation, i, prev_fitness, prev_inlier_rmse, device, dtype,
-                result);
+        std::tie(result, iteration_count) = DoSingleScaleICPIterations(
+                source_down_pyramid[scale_idx], target_down_pyramid[scale_idx],
+                target_nns, criterias[scale_idx],
+                max_correspondence_distances[scale_idx], estimation, scale_idx,
+                iteration_count, device, dtype, result,
+                callback_after_iteration);
 
         // To calculate final `fitness` and `inlier_rmse` for the current
         // `transformation` stored in `result`.
-        if (i == num_iterations - 1) {
-            GetRegistrationResultAndCorrespondences(
-                    source_down_pyramid[i], target_nns,
-                    max_correspondence_distances[i], transformation, result);
+        if (scale_idx == num_scales - 1) {
+            result = GetRegistrationResultAndCorrespondences(
+                    source_down_pyramid[scale_idx], target_nns,
+                    max_correspondence_distances[scale_idx],
+                    result.transformation_);
         }
 
         // No correspondences.
```

Comment:
Instead of `return`, use `break`. It is the clearest if a function only has one return statement, as it is easy to follow where the function terminates.

Reply:
Done.